lib/charms/observability_libs/v1/cert_handler.py

```diff
@@ -0,0 +1,391 @@
+```python
+self.framework.observe(self.cert_handler.on.cert_changed, self._on_server_cert_changed)
+
+container.push(keypath, self.cert_handler.key)
+container.push(certpath, self.cert_handler.cert)
+```
+
+This library requires a peer relation to be declared in the requirer's metadata. Peer relation data
+is used for "persistent storage" of the private key and certs.
+"""
+import ipaddress
+import json
+import socket
+from itertools import filterfalse
+from typing import List, Optional, Union
+
+try:
+    from charms.tls_certificates_interface.v2.tls_certificates import (  # type: ignore
+        AllCertificatesInvalidatedEvent,
+        CertificateAvailableEvent,
+        CertificateExpiringEvent,
+        CertificateInvalidatedEvent,
+        TLSCertificatesRequiresV2,
+        generate_csr,
+        generate_private_key,
+    )
+except ImportError as e:
+    raise ImportError(
+        "failed to import charms.tls_certificates_interface.v2.tls_certificates; "
+        "Either the library itself is missing (please get it through charmcraft fetch-lib) "
+        "or one of its dependencies is unmet."
+    ) from e
+
+import logging
+
+from ops.charm import CharmBase, RelationBrokenEvent
+from ops.framework import EventBase, EventSource, Object, ObjectEvents
+from ops.model import SecretNotFoundError
+
+logger = logging.getLogger(__name__)
+
+
+LIBID = "b5cd5cd580f3428fa5f59a8876dcbe6a"
+LIBAPI = 1
+LIBPATCH = 1
+
+
+def is_ip_address(value: str) -> bool:
+    """Return True if the input value is a valid IPv4 address; False otherwise."""
+    try:
+        ipaddress.IPv4Address(value)
+        return True
+    except ipaddress.AddressValueError:
+        return False
+
+
+class CertChanged(EventBase):
+    """Event raised when a cert is changed (becomes available or revoked)."""
+
+
+class CertHandlerEvents(ObjectEvents):
+    """Events for CertHandler."""
+
+    cert_changed = EventSource(CertChanged)
+
+
+class CertHandler(Object):
+    """A wrapper for the requirer side of the TLS Certificates charm library."""
+
+    on = CertHandlerEvents()  # pyright: ignore
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        *,
+        key: str,
+        certificates_relation_name: str = "certificates",
+        cert_subject: Optional[str] = None,
+        sans: Optional[List[str]] = None,
+    ):
+        """CertHandler is used to wrap TLS Certificates management operations for charms.
+
+        CerHandler manages one single cert.
+
+        Args:
+            charm: The owning charm.
+            key: A manually-crafted, static, unique identifier used by ops to identify events.
+             It shouldn't change between one event to another.
+            certificates_relation_name: Must match metadata.yaml.
+            cert_subject: Custom subject. Name collisions are under the caller's responsibility.
+            sans: DNS names. If none are given, use FQDN.
+        """
+        super().__init__(charm, key)
+
```

Comment:
I'd add here a `_check_juju_supports_secrets` method call that verifies that we have the required hook tools

Reply:
and raise a big exception if not